### PR TITLE
Make noise variance consistent in the unit test of batched Cov2d class

### LINKED ISSUE #307: Batched covar2d CWF coefficient resolution
There appears to be a larger than usual disagreement between the `BatchedRotCov2D` and `RotCov2D` CWF coefficients (`get_cwf_coeffs`) when changing between `float32` and `float64`.

This is currently known as:
https://github.com/ComputationalCryoEM/ASPIRE-Python/blob/09eab437033bbbe6fb296d7d3301467e03b3772f/tests/test_batched_covar2d.py#L129

On my machine I respectively get the following `max(abs(diff))`:
2.8E-4 vs 1.7E-11

Most other ASPIRE methods can achieve E-6, with a select few at E-5 in singles.  It would be good to identify where this diminished resolution is coming from. Whether it is a bug, artifact, or reasonable I am unsure at this time.

### PR DESCRIPTION
**Pull request by junchaoxia**
Closing issue #307. The test comparing the CWF coefficients from Cov2D to BatchedCov2D failed after switching dtype to float32. Here all tests are cleaned up and made more consistent to each other. In the original there is  one major reason might lead to numerical instablilty of `testCWFCoeff`. The initialized simulation objects have zero noise variance while 0.1848 is used for evaluating covariance matrices and coefficients.  In the original test, due to the mislead naming of two BatchedCov2D objects, it compared the results from two BatchedCov2D objects (one in setup and the other inside the test function)  instead of one from Cov2D and one from BatchedCov2d.

### CODE REVIEW

**Comment by janden** [comment]
Please provide a description of what this PR does. In other words, what was the underlying problem in #307 that it resolves?

**Comment by junchaoxia** [comment]
@janden  I added the more descriptions to the PR.

**Comment by janden** [comment]
> Here all tests are cleaned up and made more consistent to each other. In the original there is  one major reason might lead to numerical instablilty of `testCWFCoeff`. The initialized simulation objects have zero noise variance while 0.1848 is used for evaluating covariance matrices and coefficients. 

Right, but as we discussed, this doesn't mean that the underlying bug is fixed. I'm reluctant to change the test so that it no longer exposes the bug if that bug is still present. We need to understand what's going on and *why* the mismatch in noise variance causes the difference between the Cov2D and the BatchedCov2D classes.

> In the original test, due to the mislead naming of two BatchedCov2D objects, it compared the results from two BatchedCov2D objects (one in setup and the other inside the test function)  instead of one from Cov2D and one from BatchedCov2d.   

What was the effect of this? Did this also expose the same instability issue or is it unrelated?

**Comment by junchaoxia** [comment]
> Right, but as we discussed, this doesn't mean that the underlying bug is fixed. I'm reluctant to change the test so that it no longer exposes the bug if that bug is still present. We need to understand what's going on and why the mismatch in noise variance causes the difference between the Cov2D and the BatchedCov2D classes.

When I used noise variance 10 times smaller such as 0.01846. The  original unit test got passed. This is the way I found  the problem. I believe the level of noise do have big effect on finding covariance matrix if the numerical is not accurate enough as shown in the lines after below: 
https://github.com/ComputationalCryoEM/ASPIRE-Python/blob/master/src/aspire/covariance/covar2d.py#L284
https://github.com/ComputationalCryoEM/ASPIRE-Python/blob/master/src/aspire/covariance/covar2d.py#L572

>What was the effect of this? Did this also expose the same instability issue or is it unrelated?

The unit test  should compare the results between Cov2D and BatchedCov2D. So we have to make the correction.

**Comment by janden** [comment]
>  I believe the level of noise do have big effect on finding covariance matrix if the numerical is not accurate enough as shown in the lines after below: 
https://github.com/ComputationalCryoEM/ASPIRE-Python/blob/master/src/aspire/covariance/covar2d.py#L284
https://github.com/ComputationalCryoEM/ASPIRE-Python/blob/master/src/aspire/covariance/covar2d.py#L572

Ok but as you see those two lines *do the same thing*, so why should a mismatch in the estimated noise variance give different results? We've been over this several times. Why does f(x) = g(x) but f(y) ≠ g(y) if f and g are supposed to do the same thing?

> The unit test should compare the results between Cov2D and BatchedCov2D. So we have to make the correction.

I understand that, but what does it have to do with the reported instability in the coefficient estimates? Is it related or not?

**Comment by junchaoxia** [comment]
For further study the problem,  add more test details after setting the noise variance to zero for initializing the original images and comparing the right results from Cov2D and BatachedCov2D.

1) When we use float64,  the testCWFCoeff test can be passed without problem.
2) The test can not be passed when switch to float32 if we use large noise variance (0.1848) for building covariance matrix. But if we reduce the noise variance to 0.001848 (100 times smaller),  the test can be passed.  It looks that the noise variance for building covariance matrix can not be too far away for the value used to initialize the noisy images.      

3) The major reason for the failure at large noise variance (0.1848) is that although the differences between covariance matrices are the order of ~ 1e-6,   some differences of CWF coefficients are on the order of  ~1e-2.  

Covar_Cov2D:
[[-0.07211319 -0.1329556  -0.2469936  -0.33586374]
 [-0.13295558  0.03568286  0.51081747 -0.07518257]
 [-0.2469936   0.51081747 -0.25950116 -0.15569438]
 [-0.3358637  -0.07518257 -0.15569437 -0.18051222]]
[[ 0.3490833   0.53399897  0.38579574]
 [ 0.533999   -0.04976876  0.70820844]
 [ 0.38579583  0.70820844  0.23333327]]
[[ 0.3490833   0.53399897  0.38579574]
 [ 0.533999   -0.04976876  0.70820844]
 [ 0.38579583  0.70820844  0.23333327]]
[[-0.42562276  0.2388004  -0.0539052 ]
 [ 0.2388004  -0.26643115  0.06570511]
 [-0.05390519  0.06570511 -0.38030943]]
[[-0.42562276  0.2388004  -0.0539052 ]
 [ 0.2388004  -0.26643115  0.06570511]
 [-0.05390519  0.06570511 -0.38030943]]
[[-2.408033  -1.3820049]
 [-1.3820049 -1.6559248]]
[[-2.408033  -1.3820049]
 [-1.3820049 -1.6559248]]
[[-1.1296002  -0.48230803]
 [-0.48230806 -1.0379046 ]]
[[-1.1296002  -0.48230803]
 [-0.48230806 -1.0379046 ]]
[[-0.6485155  -0.03279277]
 [-0.03279277 -0.6853441 ]]
[[-0.6485155  -0.03279277]
 [-0.03279277 -0.6853441 ]]
[[-1.2999214]]
[[-1.2999214]]
[[-1.0435929]]
[[-1.0435929]]
[[-0.86301947]]
[[-0.86301947]]

Covar_BatchedCov2D:
[[-0.07211445 -0.1329532  -0.2469934  -0.33586165]
 [-0.1329532   0.03568358  0.51081884 -0.07518248]
 [-0.2469934   0.51081884 -0.25949773 -0.15569326]
 [-0.33586168 -0.07518246 -0.15569328 -0.18051256]]
[[ 0.34908387  0.5339991   0.38579544]
 [ 0.533999   -0.04976892  0.70820856]
 [ 0.3857954   0.70820856  0.23333329]]
[[ 0.34908387  0.5339991   0.38579544]
 [ 0.533999   -0.04976892  0.70820856]
 [ 0.3857954   0.70820856  0.23333329]]
[[-0.4256227   0.23880038 -0.05390521]
 [ 0.23880036 -0.26643118  0.0657051 ]
 [-0.05390521  0.06570511 -0.38030946]]
[[-0.4256227   0.23880038 -0.05390521]
 [ 0.23880036 -0.26643118  0.0657051 ]
 [-0.05390521  0.06570511 -0.38030946]]
[[-2.4080336 -1.3820052]
 [-1.3820052 -1.6559249]]
[[-2.4080336 -1.3820052]
 [-1.3820052 -1.6559249]]
[[-1.1296002  -0.48230803]
 [-0.48230806 -1.0379046 ]]
[[-1.1296002  -0.48230803]
 [-0.48230806 -1.0379046 ]]
[[-0.6485156  -0.03279277]
 [-0.03279277 -0.68534404]]
[[-0.6485156  -0.03279277]
 [-0.03279277 -0.68534404]]
[[-1.2999214]]
[[-1.2999214]]
[[-1.0435929]]
[[-1.0435929]]
[[-0.86301947]]
[[-0.86301947]]

Coeff_Cov2D from image 27:
[ 5.99934807e+01  1.15872755e+01  7.58538742e+01  6.33287735e+01
  2.53178358e-01  1.51281863e-01  3.71576667e-01 -1.46494389e-01
 -4.13470149e-01 -2.02656299e-01 -1.18130970e+00  6.11744225e-01
  1.77789068e+00 -1.00314081e+00  1.91859663e-01  1.01472485e+00
 -6.76015854e-01 -4.16246355e-01 -7.93085217e-01 -2.69613951e-03
  7.48590565e+00  6.93094921e+00  2.32017636e+00  2.19784069e+00
  1.68521559e+00  3.17886806e+00  5.93922377e-01  8.45888376e-01
 -1.06583335e-01  1.42705783e-01 -2.06302434e-01 -2.17589349e-01
 -1.78624794e-01  5.40637113e-02]

Coeff_BatchedCov2D from image 27:
[ 6.00070190e+01  1.15897045e+01  7.58713531e+01  6.33439941e+01
  2.53178895e-01  1.51281759e-01  3.71576279e-01 -1.46495089e-01
 -4.13470209e-01 -2.02655509e-01 -1.18130922e+00  6.11744046e-01
  1.77789056e+00 -1.00313997e+00  1.91859081e-01  1.01472402e+00
 -6.76016390e-01 -4.16246742e-01 -7.93085992e-01 -2.69682286e-03
  7.48590565e+00  6.93094921e+00  2.32017636e+00  2.19784069e+00
  1.68521643e+00  3.17886901e+00  5.93922555e-01  8.45888555e-01
 -1.06583335e-01  1.42705783e-01 -2.06302434e-01 -2.17589349e-01
 -1.78624794e-01  5.40637113e-02]

4) If we use the covariance matrix from Cov2D to calculate CWF coefficients using BatchedCov2D `get_cwf_coeffs` function, then the test can be passed even at large noise variance. So we should focus on the way for building covariance matrix to reduce the differences of CWF coefficients.

**Comment by janden** [comment]
> The test can not be passed when switch to float32 if we use large noise variance (0.1848) for building covariance matrix. But if we reduce the noise variance to 0.001848 (100 times smaller), the test can be passed. It looks that the noise variance for building covariance matrix can not be too far away for the value used to initialize the noisy images.

So the true variance in the data is zero, correct? What you're saying is that when the mismatch between the true variance and the estimated variance is big, we hit this instability?

> The major reason for the failure at large noise variance (0.1848) is that although the differences between covariance matrices are the order of ~ 1e-6, some differences of CWF coefficients are on the order of ~1e-2.

Ok the problem is somewhere in the coefficient calculation, then. From the results you've posted, it seems that the error is larger at the lower frequencies compared to the high ones. This may be a clue to what's happening.

> So we should focus on the way for building covariance matrix to reduce the differences of CWF coefficients.

I don't follow. It looks like the covariance matrix has a relatively small error. The instability is in the calculation of the CWF coefficients. So that's what we need to investigate.

**Comment by junchaoxia** [comment]
> So the true variance in the data is zero, correct? What you're saying is that when the mismatch between the true variance and the estimated variance is big, we hit this instability?

Yes. When we calculate the covariance matrix, we need to specify the noise variance (such as 0.1848) which in real case should be estimated from by a noise estimator.  

> I don't follow. It looks like the covariance matrix has a relatively small error. The instability is in the calculation of the CWF coefficients. So that's what we need to investigate.

In the four points I listed, I tried to use the covariance matrix computed from the Cov2D object to calculate CWF coefficient using the same BatchedCov2D `get_cwf_coeffs` function, then the test can be passed. It means that although the differences between covariance matrices are small (the order of ~ 1e-6), but can lead big differences of CWF coefficients (on the order of ~1e-2).  The methods for calculating the covariance matrix in Cov2D and BatchedCov2D are theoretically identical but were implemented in different ways, which leads small differences but can generate larger errors for CWF coefficients when the noise variance specified are far away to the true value from images. That's the main summary I can get.

**Comment by janden** [comment]
> which leads small differences but can generate larger errors for CWF coefficients when the noise variance specified are far away to the true value from images. 

Right, so why is that? This is what needs to be investigated and potentially corrected.

**Comment by junchaoxia** [comment]
As we discussed before, we should focus on checking posititive semidefinite property of the covariance matrix. The main problem is how to match the noise variance as mentioned before.  The posititive semidefinite property can be violated greatly when the noise variance for initializing the noise images is far away the noise variance for calculating covariance matrix.  

If we initialize the images without noise but using noise variance of 0.1848 for calculating covariance matrix. The left side B for calculating covariance matrix is positive semidenfinte before noise correction from line:
> https://github.com/ComputationalCryoEM/ASPIRE-Python/blob/master/src/aspire/covariance/covar2d.py#L697
All eigenvalues: 
[0.10803008 0.04273971 0.00235195 0.01151662 0.3011153  0.03948477
 0.00225502 0.3011153  0.03948477 0.00225502 0.03034246 0.00170675
 0.01356218 0.03034246 0.00170675 0.01356218 0.00082276 0.00152385
 0.00082276 0.00152385 0.00103117 0.00390697 0.00103117 0.00390697
 0.00589821 0.00095731 0.00589821 0.00095731 0.00077885 0.00077885
 0.0006485  0.0006485  0.00130312 0.00130312]

It becomes not after noise correction from line:
>  https://github.com/ComputationalCryoEM/ASPIRE-Python/blob/master/src/aspire/covariance/covar2d.py#L698
All eigenvalues: 
[ 0.03846596  0.00113658 -0.03836413 -0.0234469   0.21915819  0.01690725
 -0.03620818  0.21915819  0.01690725 -0.03620818 -0.05278548 -0.01198728
 -0.03305352 -0.05278548 -0.01198728 -0.03305352 -0.00637515 -0.03562982
 -0.00637515 -0.03562982 -0.03382645 -0.01102123 -0.03382645 -0.01102123
 -0.03200343 -0.02252299 -0.03200343 -0.02252299 -0.0149834  -0.0149834
 -0.01904513 -0.01904513 -0.0209753  -0.0209753 ]

The resulted covariance matrix has many eigenvalues with large negative values(30 out of 34): 
> https://github.com/ComputationalCryoEM/ASPIRE-Python/blob/master/src/aspire/covariance/covar2d.py#L702
All eigenvalues: 
[ 0.529986    0.18924311 -0.738403   -0.45726886  1.2644975  -0.06332816
 -0.66852164  1.2644975  -0.06332816 -0.66852164 -0.09258118 -0.62538064
 -0.3544015  -0.09258118 -0.62538064 -0.3544015  -3.4642344  -0.5997239
 -3.4642344  -0.5997239  -1.5682346  -0.59927016 -1.5682346  -0.59927016
 -0.6293206  -0.704539   -0.6293206  -0.704539   -1.2999214  -1.2999214
 -1.0435929  -1.0435929  -0.86301947 -0.86301947]

If we initialize images with noise variance of 0.1848. Then number of eigenvalues with negative values can be reduced greatly (14/34) and so are the magnitudes, as shown below:
> All eigenvalues: 
 [ 1.0591569   0.5113819  -0.40236154 -0.13609916 -0.25375184  0.9590567
  1.1974989  -0.25375184  0.9590567   1.1974989   0.40656665  0.15562901
  0.03423601  0.40656665  0.15562901  0.03423601  0.36734086 -0.22970942
  0.36734086 -0.22970942  0.51550037  0.03141885  0.51550037  0.03141885
  0.14809038 -0.0769847   0.14809038 -0.0769847  -0.04749447 -0.04749447
 -0.3905021  -0.3905021  -0.10768896 -0.10768896]

 I have added the function to check positive semidefinite property and generate warning when necessary in commit 5b8670a.

**Comment by janden** [comment]
Thank you for looking over this. This makes a lot of sense. I think a reasonable solution is to ensure that the covariance matrix is positive semidefinite after the CG step. Could you add this? It would be good to add a flag that's true by default so that users can turn if off if needed. I suspect this type of correction is only necessary for very small datasets.

**Comment by junchaoxia** [comment]
@janden To make it positive semidefinite, is it good enough to just replacing the non-diagonal elements with an average, (c_ij+c_ji)/2?

**Comment by junchaoxia** [comment]
It looks that the original one is quite symetrical. Need to find the nearest SPD as discussed in the link below: 
https://stackoverflow.com/questions/43238173/python-convert-matrix-to-positive-semi-definite/43244194

**Comment by janden** [comment]
Simplest way to get the closest (in Frobenius norm) positive semidefinite matrix is to set all the negative eigenvalues to zero.

**Comment by junchaoxia** [comment]
After setting the negative eigenvalues to zero, the covariance matrix with zero noise variance can be rebuilt as below:  
[[ 0.16145925 -0.11753237 -0.10274341 -0.08353995]
 [-0.11753234  0.27855277  0.22699049 -0.03799514]
 [-0.10274342  0.22699057  0.18540677 -0.02476057]
 [-0.08353996 -0.03799509 -0.02476053  0.09380959]]
[[0.42056045 0.405879   0.43610638]
 [0.40587908 0.39171013 0.4208823 ]
 [0.43610635 0.4208822  0.45222694]]
[[0.42056045 0.405879   0.43610638]
 [0.40587908 0.39171013 0.4208823 ]
 [0.43610635 0.4208822  0.45222694]]
[[0. 0. 0.]
 [0. 0. 0.]
 [0. 0. 0.]]
[[0. 0. 0.]
 [0. 0. 0.]
 [0. 0. 0.]]
[[0. 0.]
 [0. 0.]]
[[0. 0.]
 [0. 0.]]
[[0. 0.]
 [0. 0.]]
[[0. 0.]
 [0. 0.]]
[[0. 0.]
 [0. 0.]]
[[0. 0.]
 [0. 0.]]
[[0.]]
[[0.]]
[[0.]]
[[0.]]
[[0.]]
[[0.]]

Both tests for covarince matrices and CWF coefficients can be passed with the `atol=1e-6` even using the initial images without noises. Add the option to rebuild positive definite covariance in commit 76c6e53.

File: src/aspire/utils/matrix.py [hunk 1]
```diff
@@ -299,6 +299,21 @@ def make_symmat(A):
...
+
+    This is the simplest way by setting negative eigenvalues to zero.
+    :param A: A matrix.
+    :return: The positive semi-definite matrix
+    """
+    B = make_symmat(A)
+    W, V = eigh(B)
+    W[W < 0.0] = 0.0
+
+    return V @ np.diag(W) @ np.linalg.inv(V)
```
Comment (janden) [inline src/aspire/utils/matrix.py]: `np.linalg.inv(V)` → `np.transpose(V)`

File: src/aspire/utils/matrix.py [hunk 2]
```diff
@@ -299,6 +299,21 @@ def make_symmat(A):
     return 0.5 * (A + A.T)
 
 
+def psd_mat(A):
```
Comment (janden) [inline src/aspire/utils/matrix.py]: Just call it `make_psd` to align with `make_symmat`.

File: src/aspire/covariance/covar2d.py [hunk 3]
```diff
@@ -208,6 +208,7 @@ def get_covar(
         do_refl=True,
         noise_var=1,
         covar_est_opt=None,
+        make_psd=False,
```
Comment (janden) [inline src/aspire/covariance/covar2d.py]: Default to `True`.

File: src/aspire/operators/blk_diag_matrix.py [hunk 4]
```diff
@@ -679,6 +680,21 @@ def check_psd_matrix(self):
         )
         return np.alltrue(eigenvalues > 0.0)
 
+    def make_psd_matrix(self):
```
Comment (janden) [inline src/aspire/operators/blk_diag_matrix.py]: Just `make_psd` is enough.

File: src/aspire/operators/blk_diag_matrix.py [hunk 5]
```diff
@@ -679,6 +680,21 @@ def check_psd_matrix(self):
...
+        """
+        Convert all submatrices to positive semidefinite
```
Comment (janden) [inline src/aspire/operators/blk_diag_matrix.py]: “submatrices” → “blocks”

File: src/aspire/operators/blk_diag_matrix.py [hunk 6]
```diff
@@ -679,6 +680,21 @@ def check_psd_matrix(self):
...
+
+        :return: The BlkDiagMatrix instance with all matrices of
```
Comment (janden) [inline src/aspire/operators/blk_diag_matrix.py]: “with all blocks positive semidefinite”

File: src/aspire/operators/blk_diag_matrix.py [hunk 7]
```diff
@@ -668,6 +668,17 @@ def apply(self, X):
 
         return Y
 
+    def check_psd_matrix(self):
```
Comment (janden) [inline src/aspire/operators/blk_diag_matrix.py]: Just `check_psd`.

File: src/aspire/operators/blk_diag_matrix.py [hunk 8]
```diff
@@ -668,6 +668,17 @@ def apply(self, X):
...
+        """
+        Check the positive semidefinite property of all submatrices
```
Comment (janden) [inline src/aspire/operators/blk_diag_matrix.py]: “submatrices” → “blocks”

File: src/aspire/operators/blk_diag_matrix.py [hunk 9]
```diff
@@ -668,6 +668,17 @@ def apply(self, X):
...
+
+        :return: True if all matrices have non-negative eigenvalues.
```
Comment (janden) [inline src/aspire/operators/blk_diag_matrix.py]: “matrices” → “blocks”

File: src/aspire/covariance/covar2d.py [hunk 10]
```diff
@@ -281,6 +281,10 @@ def identity(x):
             A[k] = np.sqrt(weight) * ctf_fb_k_sq
             M += A[k]
 
+        logger.info("Check positive semidefinite property of left side b in Cov2D.")
```
Comment (janden) [inline src/aspire/covariance/covar2d.py]: Not needed, remove. (Same for similar `logger.info` below.)

File: src/aspire/covariance/covar2d.py [hunk 11]
```diff
@@ -281,6 +281,10 @@ def identity(x):
...
+        if not b_coeff.check_psd_matrix():
+            logger.warning("Matrices are not positive semidefinite.")
```
Comment (janden) [inline src/aspire/covariance/covar2d.py]: Which matrix? Be specific. (Same below.)

**Review by janden — CHANGES_REQUESTED** [review]

File: src/aspire/utils/matrix.py (hunk 1 above)
Comment (junchaoxia) [reply src/aspire/utils/matrix.py]: Done in new commit 8d08e36.

File: src/aspire/utils/matrix.py (hunk 2 above)
Comment (junchaoxia) [reply src/aspire/utils/matrix.py]: Changed to `make_psd` in new commit 8d08e36.

File: src/aspire/operators/blk_diag_matrix.py (hunk 4 above)
Comment (junchaoxia) [reply src/aspire/operators/blk_diag_matrix.py]: Changed to `make_psd` in new commit 8d08e36.

File: src/aspire/operators/blk_diag_matrix.py (hunk 5 above)
Comment (junchaoxia) [reply src/aspire/operators/blk_diag_matrix.py]: Done in new commit 8d08e36.

File: src/aspire/operators/blk_diag_matrix.py (hunk 6 above)
Comment (junchaoxia) [reply src/aspire/operators/blk_diag_matrix.py]: Done in new commit 8d08e36.

File: src/aspire/operators/blk_diag_matrix.py (hunk 7 above)
Comment (junchaoxia) [reply src/aspire/operators/blk_diag_matrix.py]: Done in new commit 8d08e36.

File: src/aspire/operators/blk_diag_matrix.py (hunk 8 above)
Comment (junchaoxia) [reply src/aspire/operators/blk_diag_matrix.py]: Done in new commit 8d08e36.

File: src/aspire/operators/blk_diag_matrix.py (hunk 9 above)
Comment (junchaoxia) [reply src/aspire/operators/blk_diag_matrix.py]: Done in new commit 8d08e36.

File: src/aspire/covariance/covar2d.py (hunk 10 above)
Comment (junchaoxia) [reply src/aspire/covariance/covar2d.py]: Done in new commit 8d08e36.

File: src/aspire/covariance/covar2d.py (hunk 11 above)
Comment (junchaoxia) [reply src/aspire/covariance/covar2d.py]: Done in new commit 8d08e36.

File: src/aspire/covariance/covar2d.py (hunk 3 above)
Comment (junchaoxia) [reply src/aspire/covariance/covar2d.py]: If we set the default to `True`, we need to change the related unit tests for the original Cov2D. Please let me know your preference.

Comment (janden) [reply src/aspire/covariance/covar2d.py]: Then we'll have to change the unit tests. It doesn't make sense to output a non-PSD covariance matrix by default.

**Comment by garrettwrong** [comment]
Closes #329

File: src/aspire/covariance/covar2d.py (hunk 3 above)
Comment (garrettwrong) [reply src/aspire/covariance/covar2d.py]: I have updated the default to True and regenerated three corresponding unit test files.

**Comment by garrettwrong** [comment]
Just noting I accidentally forgot to pull branch this before I started to work, so I had to add back 8d08e36f14adc0386 which was more recent than my prior checkout.

Local tests passed. Assuming CI passes I'll request the review to close this out.

File: src/aspire/covariance/covar2d.py (hunk 3 above)
Comment (janden) [reply src/aspire/covariance/covar2d.py]: Thanks!

**Review by janden — APPROVED** [review]
